#!/usr/bin/env python
PACKAGE = "limo_mission"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# HLS thresholds for lane detection (WHITE lanes on BLACK background)
# H = Hue, L = Lightness, S = Saturation
gen.add("lane_h_low", int_t, 0, "Lane H channel low threshold", 0, 0, 180)
gen.add("lane_h_high", int_t, 0, "Lane H channel high threshold", 180, 0, 180)
gen.add("lane_l_low", int_t, 0, "Lane L channel low threshold", 180, 0, 255)
gen.add("lane_l_high", int_t, 0, "Lane L channel high threshold", 255, 0, 255)
gen.add("lane_s_low", int_t, 0, "Lane S channel low threshold", 0, 0, 255)
gen.add("lane_s_high", int_t, 0, "Lane S channel high threshold", 60, 0, 255)

# Dual ROI settings for curve detection
gen.add("roi_top", int_t, 0, "Far ROI top position (look-ahead)", 280, 0, 480)
gen.add("roi_mid", int_t, 0, "ROI split point (far/near boundary)", 380, 0, 480)
gen.add("roi_bottom", int_t, 0, "Near ROI bottom position", 480, 0, 480)

# PD Control parameters
gen.add("base_speed", double_t, 0, "Base driving speed", 0.3, 0.0, 1.0)
gen.add("kp", double_t, 0, "Proportional gain for steering", 0.008, 0.0, 0.05)
gen.add("kd", double_t, 0, "Derivative gain for smoother response", 0.004, 0.0, 0.05)
gen.add("max_steering", double_t, 0, "Maximum steering angle", 0.5, 0.0, 1.0)

# Curve handling parameters
gen.add("lookahead_weight", double_t, 0, "Weight of far ROI for curve prediction (0.0-0.7)", 0.3, 0.0, 0.7)
gen.add("curve_speed_factor", double_t, 0, "Speed reduction factor in curves (0.3-0.8)", 0.5, 0.0, 1.0)
gen.add("lane_offset", int_t, 0, "Offset from lane when only one lane visible (pixels)", 150, 50, 300)

exit(gen.generate(PACKAGE, "lane_detect_node", "LaneDetect"))
